### DISCUSSION TITLE
Make case statements define variable scopes

### DISCUSSION THREAD

**Discussion by TahirAhmadov** [dotnet/csharplang#5636] [Language Ideas] [answered]
I've never understood why this is illegal in C#. Can we change it?
```cs
switch(x)
{
  case 1:
    string a = GetString1();
    ProcessString(a);
    break;
  case 2:
    string a = GetString2();
    ProcessString(a);
    break;
}
```
However, this is legal:
```cs
switch(x)
{
  case 1:
    {
      string a = GetString1();
      ProcessString(a);
      break;
    }
  case 2:
    {
      string a = GetString2();
      ProcessString(a);
      break;
    }
}
```

**Comment by HaloFour** [comment, marked as answer]
C# inherits the behavior from C, for better or for worse.  Cases are labels and labels don't introduce scope.  It can't be changed as it would be a massive breaking change.

The second example works because you're using blocks, which do introduce scope.

**Reply by TahirAhmadov** [reply]
I understand this was inherited from C, but C had a different design - they allowed to drop from one case to another (IIRC). Which brings me to my second point:
Cases are labels in _implementation_, not in semantics. In C# semantics, they are already blocks.

**Reply by HaloFour** [reply]
Yes, C# requires explicit fallthrough because that was a well known source of bugs in C and Java.  That was a design decision that was made more than 20 years ago.  Variable scoping doesn't lead to those kinds of bugs, at worst it leads to compile-time errors.

Either way, it's 20 years too late to consider fundamental changes to how this language structure behaves.

**Reply by TahirAhmadov** [reply]
> Yes, C# requires explicit fallthrough because that was a well known source of bugs in C and Java. That was a design decision that was made more than 20 years ago. 

I think you meant "prohibits", not "requires". For the record, I'm fully supportive of that decision.

> Variable scoping doesn't lead to those kinds of bugs, at worst it leads to compile-time errors.
> Either way, it's 20 years too late to consider fundamental changes to how this language structure behaves.

Yes, I wasn't arguing that variable scoping introduces bugs; it is just annoying compilation errors.
WRT it being too late to change now, I would disagree. Treating cases as blocks would introduce no new bugs whatsoever; something that was illegal before becomes legal now, but nothing that was legal before starts working differently now. It's purely a lifting of a silly restriction with no side effects whatsoever.

**Reply by 333fred** [reply]
> I think you meant "prohibits", not "requires". For the record, I'm fully supportive of that decision.

No, Halo meant requires. There is no implicit fall through (that's the thing we prohibit). If you want fall through, you must use goto (ie explicit fall through).

**Reply by HaloFour** [reply]
> I think you meant "prohibits", not "requires".

I said "requires explicit fallthrough", in that you are required to explicitly specify that you want to fallthrough between cases with `goto` rather than it occurring implicitly.

> Treating cases as blocks would introduce no new bugs whatsoever

It causes code that expects the scope to span the labels to break.

> it is just annoying compilation errors.

Which are easily fixed by including a block in the case label.

**Reply by TahirAhmadov** [reply]
Oh OK, I see now what he meant, I didn't notice the word "explicit". Still, I think I understood what he meant correctly overall.

**Reply by TahirAhmadov** [reply]
> > Treating cases as blocks would introduce no new bugs whatsoever
> 
> It causes code that expects the scope to span the labels to break.

Scope doesn't span labels though. You can't access that variable `a` from second case if it was defined in the first case.
 
> > it is just annoying compilation errors.
> 
> Which are easily fixed by including a block in the case label.

Yes, but lifting this error (aka making cases "blocks" for scoping) is also very easy.

**Reply by HaloFour** [reply]
> Scope doesn't span labels though. You can't access that variable a from second case if it was defined in the first case.

Without blocks the variable is scoped across all case labels in the switch statement.

> Yes, but lifting this error (aka making cases "blocks" for scoping) is also very easy.

Changing the meaning of 20 year old code definitely does not qualify as "very easy".

**Reply by TahirAhmadov** [reply]
> Without blocks the variable is scoped across all case labels in the switch statement.

It is in terms of prohibiting the redefinition of another variable, but not in terms of being able to access the first variable. That's exactly the problem, it's scoped "negatively" but not "positively", which is pointless.

**Reply by HaloFour** [reply]
> It is in terms of prohibiting the redefinition of another variable, but not in terms of being able to access the first variable.

Sure you can, depending on definite assignment rules.  The scope is the entire `switch`, but the compiler does track and enforce that you don't read the variable unless it can ensure that it has been definitely assigned.

**Reply by bernd5** [reply]
What Halo wants to say is that you could write to "a" in the 2. case without a separate declaration.
This would break...

**Reply by TahirAhmadov** [reply]
> Sure you can

Can you show a code example?

**Reply by bernd5** [reply]
Just remove the type (string) from your example in the second case...

**Reply by TahirAhmadov** [reply]
> What Halo wants to say is that you could write to "a" in the 2. case without a separate declaration. This would break...

Oh my God, you are right. I never knew that was the case. Consider the question closed.
